application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,129 @@
+            .flatMap(event::replyEmbeds);
+    }
+
+    private static @NotNull RestAction<Boolean> dmUser(@NotNull ISnowflake target,
+            @NotNull String reason, @NotNull Guild guild, @NotNull SlashCommandEvent event) {
+        return event.getJDA()
+            .openPrivateChannelById(target.getId())
+            .flatMap(channel -> channel.sendMessage(
+                    """
+                            Hey there, sorry to tell you but unfortunately you have been warned in the server %s.
+                            If you think this was a mistake, please contact a moderator or admin of the server.
+                            The reason for the warning is: %s
+                            """
+                        .formatted(guild.getName(), reason)))
+            .mapToResult()
+            .map(Result::isSuccess);
+    }
+
+    private void warnUser(@NotNull User target, @NotNull Member author, @NotNull String reason,
+            @NotNull Guild guild) {
+        logger.info("'{}' ({}) warned the user '{}' ({}) for reason '{}'.",
+                author.getUser().getAsTag(), author.getId(), target.getAsTag(), target.getId(),
+                reason);
+
+        actionsStore.addAction(guild.getIdLong(), author.getIdLong(), target.getIdLong(),
+                ModerationUtils.Action.WARN, null, reason);
+    }
+
+    private static @NotNull MessageEmbed sendFeedback(boolean hasSentDm, @NotNull User target,
+            @NotNull Member author, @NotNull String reason) {
+        String dmNoticeText = "";
+        if (!hasSentDm) {
+            dmNoticeText = "(Unable to send them a DM.)";
+        }
+        return ModerationUtils.createActionResponse(author.getUser(), ModerationUtils.Action.WARN,
+                target, dmNoticeText, reason);
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping targetOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
+        Member author = Objects.requireNonNull(event.getMember(), "The author is null");
+        Guild guild = Objects.requireNonNull(event.getGuild(), "The guild is null");
+        String reason = Objects.requireNonNull(event.getOption(REASON_OPTION), "The reason is null")
+            .getAsString();
+
+        if (!handleChecks(guild.getSelfMember(), author, targetOption.getAsMember(), reason,
+                event)) {
+            return;
+        }
+
+        warnUserFlow(targetOption.getAsUser(), author, reason, guild, event).queue();
+    }
+
+    @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @Nullable Member target, String reason, @NotNull SlashCommandEvent event) {
+        if (target != null && !ModerationUtils.handleCanInteractWithTarget(ACTION_VERB, bot, author,
+                target, event)) {
+            return false;
+        }
```

Comment:
shouldn't this be ` target == null`  and `||` instead `&&` ?

Can't really read this code, don't like the style we tend to go to with our checks, so not sure

Reply:
> Can't really read this code, don't like the style we tend to go to with our checks, so not sure

Its the same as in `BanCommand` and `AuditCommand`. The `handleCanInteractWithTarget` requires that you give it a valid `Member` instance. So it must only be called if `target != null`.

But the method should continue if its `null` regardless and do the second check, which also works with just `User` (in case you want to warn an already banned user).

I agree that the `if` reads difficult, in particular the logic with the booleans. The `handleCanInteractWithTarget` returns `true` if the user can interact. In which case the surrounding method is supposed to return `false` to indicate that "everything is good".

So the boolean logic is reversed. The surrounding method returns whether the `warn` should be aborted or not (not the other way around).

**tl;dr:** The logic is correct and is also streamlined with the other moderation commands. Adjusting it here should also lead to adjusting it at the other places. Which is out of the scope of this PR. But I agree that it should be improved some way or another (but in a different issue+PR).

Reply:
Yeah, didn't mean it in a sense of "Change it in this PR"

And if it works, I'll resolve this, was just a check to make sure.